bootcamp_module08/contrib/student_02.py

```diff
@@ -0,0 +1,21 @@
+def make_complementary_DNA_strand(strand_in):
+    """Complements the DNA strand inputted.
+
+    Parameters
+    ----------
+    strand_in : str
+        The strand to complement
+
+    Returns
+    -------
+    str
+        The complementary DNA strand to the inputted strand_in
+
+    """
+    base_pair_complements = {"A": "T", "C": "G", "G": "C", "T": "A"}
+    complementary_strand_out = ""
+    for i, base in enumerate(strand_in.upper()):
+        if base not in base_pair_complements:
+            return "Unrecognized base {} at position {} (first base is position 0)".format(base, i)
+        complementary_strand_out += base_pair_complements[base]

```

Comment:
You could probably see some speed-up if you take a slightly different approach (especially on long string):
If you initialize `complementary_strand_out` as a list, then call `''.join()` on it.

E.g.,
```python
from random import choice


alphabet = ['a', 'c', 'g', 't']


def random_string(n):
    return ''.join(choice(alphabet) for _ in range(n))


def create_string_append(string):
    new_string = ""
    for char in string:
        new_string += char
    return new_string


def create_string_mutate(string):
    new_string = [None] * len(string)
    for i, char in enumerate(string):
        new_string[i] = char
    return ''.join(new_string)


N = int(2e6)
demo_string = random_string(N)

print('append to string')
%timeit create_string_append(demo_string)
print('mutate list')
%timeit create_string_mutate(demo_string)
```

Gives me:
```
append to string
295 ms ± 5.82 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)
mutate list
164 ms ± 2.62 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)

```